packages/cli/src/main/kotlin/elide/tool/cli/AbstractSubcommand.kt

```diff
@@ -425,92 +433,28 @@ import org.graalvm.polyglot.Engine as VMEngine
+   *
+   * The first invocation of this method will cause the [engine] to be initialized, triggering the
+   * [configureEngine] event.
    */
-  @Suppress("DEPRECATION")
-  protected open fun withVM(
-    context: ToolContext<State>,
-    userBundles: List<URI>,
-    systemBundles: List<URI>,
-    hostIO: Boolean,
-    contextBuilder: (VMContext.Builder) -> Unit = {},
-    op: VMCallable<State>,
-  ) {
-    require(!engineInitialized.get()) {
-      "Cannot re-initialize CLI guest VM"
-    }
-    engineInitialized.set(true)
-
-    logging.debug("Acquiring VM context for CLI tool")
-    val wrappedBuilder: (VMContext.Builder) -> Unit = {
-      // configure the VM as normal
-      contextBuilder.invoke(it)
-
-      // if we have a virtualized FS, mount it
-      if (userBundles.isNotEmpty() && !hostIO) {
-        val bundles = userBundles.map { fsBundleUri ->
-          // check the bundle URI
-          if (fsBundleUri.scheme == "classpath:") {
-            logging.debug("Rejecting `classpath:`-prefixed bundle: not supported by CLI")
-            throw ShellError.BUNDLE_NOT_FOUND.asError()
-          } else {
-            // make sure the file can be read
-            val file = try {
-              logging.trace("Checking bundle at URI '$fsBundleUri'")
-              File(fsBundleUri)
-            } catch (err: IOException) {
-              throw ShellError.BUNDLE_NOT_FOUND.asError()
-            }
-            logging.trace("Checking existence of '$fsBundleUri'")
-            if (!file.exists()) throw ShellError.BUNDLE_NOT_FOUND.asError()
-            logging.trace("Checking readability of '$fsBundleUri'")
-            if (!file.canRead()) throw ShellError.BUNDLE_NOT_ALLOWED.asError()
-            logging.debug("Mounting guest filesystem at URI: '$fsBundleUri'")
-            fsBundleUri
-          }
-        }
-
-        it.fileSystem(
-          EmbeddedGuestVFS.forBundle(
-            *systemBundles.plus(bundles).toTypedArray(),
-          ),
-        )
-      } else if (systemBundles.isNotEmpty() && !hostIO) {
-        logging.debug { "No user bundles, but ${systemBundles.size} system bundles present; mounting embedded" }
-        it.fileSystem(
-          EmbeddedGuestVFS.forBundle(
-            *systemBundles.toTypedArray(),
-          ),
-        )
-      } else if (hostIO) {
-        // if we're doing host I/O, mount that instead
-        logging.debug("Command-line flags indicate host I/O; mounting host filesystem")
-        it.fileSystem(HostVFS.acquire())
-      }
-    }
-    vmContextFactory.acquire(wrappedBuilder) {
-      try {
-        enter()
-        op.invoke(context, this)
-      } catch (err: AbstractToolError) {
-        // it's a known tool error. re-throw.
-        throw err
-      } catch (err: Throwable) {
-        logging.error(
+  protected open fun withContext(block: (PolyglotContext) -> Unit) {
+    logging.debug("Acquiring context for CLI tool")
```

Comment:
VFS configuration (handling the `userBundles` and `systemBundles` arguments), and the rest of the context customization logic has been moved down to subclasses, using the `configureEngine` event (see `ToolShellCommand.configureEngine` for example).